### DISCUSSION TITLE
Migrate DLR repo here into a monorepo

### DISCUSSION THREAD

**Discussion by Lamparter** [IronLanguages/ironpython3#1814] [Polls]
Hi!
Submodules are hard to manage, especially for those who don't have write access to the DLR repo and would like to make changes there as well (like me!).
There are two ways to mitigate this:
- publish DLR on NuGet (which I think it already is!) and exclusively use the NuGet version of it in this repo rather than the submodule
- migrate the DLR here and use as a monorepo
- _don't_ do either of these things and leave contributors in a pickle when they update things in DLR directly from this repo

Hopefully there's someone else here who understands what I mean 😅
Thanks!

---

I've attached a GitHub poll out of interest, though of course I don't have any authority to make this change.

**Comment by slozier** [comment]
It used to be part of the repo but was split out so it would work with both IronPython 2 and 3 (as well as IronRuby which never went anywhere). Unfortunately IronPython 2 is not yet dead (almost) so at the very least I would wait until that repo was archived to even consider it. In theory there are other projects that could be using the DLR as well (though which ones exactly I couldn't say).

Personally my interest in the DLR codebase does not go beyond supporting IronPython so leaving it separate increases the chances that someone else might take it over. :smile:

**Reply by Lamparter** [reply]
I thought a while about forking the DLR but decided it would be much nicer to contribute to the original repo, so I'll probably make a few PRs in there and see what might work. My main complaint was about the submodule itself, in that it's far far more convenient to use a NuGet package, simply because of the obvious inconvenience of a submodule which is capable of driving anybody nuts...
IronRuby sounds interesting! Unfortunately I don't know the slightest bit of Ruby in order to contribute to it in any way 😁

**Reply by Lamparter** [reply]
What I really want to do is improve the quality of the code and enhance it to be more modern; I'd like to rewrite the website from scratch as well, and improve the documentation since I want more people to be able to contribute to it - I can't believe how so many people _haven't_ heard about IronPython, or that not many people want to contribute to it.

**Reply by slozier** [reply]
Unfortunately a NuGet package is not super practical since often times debugging IronPython issues involves stepping through and modifying the DLR source code.

I do understand that submodules are a bit of a pain initially, but once you understand the concept they're not that hard to work with (at least for me).

**Reply by BCSharp** [reply]
I have the same experience as @slozier. What might be helpful though is a wiki page with some tips, best practices, and guidelines. Every time I want to work on a feature that spans DLR and IronPython (which doesn't happen frequently enough to remember the workflow) I feel like I need to figure it out all over again.

**Reply by Lamparter** [reply]
When I meant the submodule was annoying, I mostly meant that it kept crashing Git and would cause me to need to reclone the repository if an error occured.
I see how it's not practical though and obviously it's not my call at all. 😅

**Comment by Lamparter** [comment]
> Personally my interest in the DLR codebase does not go beyond supporting IronPython so leaving it separate increases the chances that someone else might take it over. 😄

Little update: I [forked](https://github.com/RiversideValley/Dynamic) the DLR and I'm conducting some experiments on it - restructuring of the project specifically as well as some other things. Hopefully the fork goes well so someone else can appreciate how amazing it is 😁
